openhtf/io/frontend/__main__.py

```diff
@@ -32,97 +32,223 @@
-  state = query_framework(name)
-  print('HACK 1')
-  if ('prompt' not in state) or (state['prompt'] == 'None'):
-    return 'NO_PROMPT'
-  return state['prompt']['id']
-
-
-@app.route('/station/<name>/prompt/<id>/', methods=['POST'])
-def prompt(name, id):
-  """Handle requests that respond to prompts."""
-  msg = json.JSONEncoder().encode({'id': id, 'response': flask.request.data})
-  query_framework(name, method='POST', message=msg)
-  return 'OK'
+gflags.DEFINE_integer('discovery_interval_s', 5,
+                      'Seconds to wait between station discovery attempts.')
+gflags.DEFINE_string('discovery_address',
+                     multicast.DEFAULT_ADDRESS,
+                     'Multicast address to ping for station discovery.')
+gflags.DEFINE_integer('discovery_port',
+                      multicast.DEFAULT_PORT,
+                      'Multicast port to use for station discovery.')
+gflags.DEFINE_integer('discovery_ttl',
+                      multicast.DEFAULT_TTL,
+                      'TTL for station discovery pings.')
+
+
+_LOG = logging.getLogger(__name__)
+
+UNKNOWN_STATION_ID = 'UNKNOWN_STATION'
+
+
+class Station(mutablerecords.Record(  # pylint: disable=too-few-public-methods,no-init
+    'Station', [], {'station_id': UNKNOWN_STATION_ID, 'state': None})):
+  """Represents a station seen on the local network.
+
+  Attributes:
+    station_id: Station's ID string.
+    state: A dictionary representation of station state with two keys:
+      framework: A dictionary representation of a TestExecutor object.
+      test: A dictionary representation of a TestState object.
+    """
+
+
+class StationStore(threading.Thread):
+  """Self-updating store of stations visible on the local network(s).
+
+  Station data is stored in the 'stations' attribute, a dictionary mapping
+  tuples of (host_ip, port) to Station records.
+  """
+  def __init__(self):
+    super(StationStore, self).__init__()
+    self._live = False
+    self.hostname = socket.gethostname()
+    self.stations = {}
+
+  @staticmethod
+  def MessageFramework(host, port, method='GET', message=None):
+    """Interact with a running instance of openhtf via its HTTP API.
+
+    The default method, 'GET', returns a JSON representation of station state,
+    while 'POST' can be used to respond to prompts. For example:
+
+      # Get the state of the openhtf instance at 192.168.2.25:10500.
+      station_state = MessageFramework(192.168.2.25, 10500)
+
+      # Respond to a prompt for the DUT's serial number:
+      MessageFramework(192.168.2.25, 10500, method='POST', 'SN-NCC1701D')
+
+    Args:
+      host: The IP address of the host running openhtf.
+      port: The TCP port of openhtf's HTTP API on host.
+      method: 'GET' or 'POST'.
+      message: Prompt response for POSTs. Ignored for GETs.
+    """
+    result = None
+    conn = httplib.HTTPConnection(host, int(port))
+    args = [method, '/']
+    if message and method == 'GET':
```

Comment:
exposing 'method' here seems wrong, MessageFramework should just have reads and writes (with responses). Two functions, strawman-named ReadFromFramework and SendToFramework or CallToFramework that take the right number of args wouldn't even share much code. Maybe a shared function or contextmanager to do the connection creation (and later cache it) and close it when the context ends?